supportedBackends/tes/src/main/scala/cromwell/backend/impl/tes/TesAsyncBackendJobExecutionActor.scala

```diff
@@ -415,36 +477,83 @@ class TesAsyncBackendJobExecutionActor(override val standardParams: StandardAsyn
       }
     } yield statusWithLog
 
-  private def queryStatusAsync(handle: StandardAsyncPendingExecutionHandle): Future[TesRunStatus] =
-    makeRequest[MinimalTaskView](HttpRequest(uri = s"$tesEndpoint/${handle.pendingJob.jobId}?view=MINIMAL")) map {
-      response =>
-        val state = response.state
-        state match {
-          case s if s.contains("COMPLETE") =>
-            jobLogger.info(s"Job ${handle.pendingJob.jobId} is complete")
-            Complete
+  private def queryStatusAndCostData(handle: StandardAsyncPendingExecutionHandle,
+                                     fetchCostData: Boolean
+  ): Future[TesRunStatus] =
+    if (fetchCostData) {
+      makeRequest[Task](HttpRequest(uri = s"$tesEndpoint/${handle.pendingJob.jobId}?view=FULL")) map { response =>
+        val fetchedCostData = for {
+          responseLogs <- response.logs
+          startTime <- responseLogs.head.start_time
+          vmCost <- responseLogs.head.metadata.flatMap(_.get("vm_price_per_hour_usd"))
+        } yield (startTime, vmCost)
+
+        fetchedCostData match {
+          case Some(data) =>
+            val costData = Option(TesVmCostData(data._1, data._2))
+            val state = response.state
+            val metadata = Map(
+              CallMetadataKeys.TaskStartTime -> data._1,
+              CallMetadataKeys.VmCostUsd -> data._2
+            )
+            tellMetadata(metadata)
+            getTesStatus(state, costData, handle.pendingJob.jobId)
+          case None => getTesStatus(response.state, Option.empty, handle.pendingJob.jobId)
+        }
+      }
+    } else {
+      makeRequest[MinimalTaskView](HttpRequest(uri = s"$tesEndpoint/${handle.pendingJob.jobId}?view=MINIMAL")) map {
+        response =>
+          val state = response.state
+          getTesStatus(Option(state), Option.empty, handle.pendingJob.jobId)
+      }
+    }
+
+  private def getTesStatus(state: Option[String], withCostData: Option[TesVmCostData], jobId: String): TesRunStatus = {
+    var data: Option[TesVmCostData] = Option.empty
+    if (withCostData.isDefined) {
+      data = withCostData
+      for {
+        costData <- withCostData
+        vmCost = costData.vmCost
+      } yield vmCost
+    }
+    state match {
+      case s if s.contains("COMPLETE") =>
+        jobLogger.info(s"Job ${jobId} is complete")
+        Complete(data)
 
-          case s if s.contains("CANCELED") =>
-            jobLogger.info(s"Job ${handle.pendingJob.jobId} was canceled")
-            Cancelled
+      case s if s.contains("CANCELED") =>
+        jobLogger.info(s"Job ${jobId} was canceled")
+        Cancelled(data)
 
-          case s if s.contains("EXECUTOR_ERROR") =>
-            jobLogger.info(s"TES reported a failure for Job ${handle.pendingJob.jobId}: '$s'")
-            Failed()
+      case s if s.contains("EXECUTOR_ERROR") =>
+        jobLogger.info(s"TES reported a failure for Job ${jobId}: '$s'")
+        Failed()
```

Comment:
Do we want to track cost data for failed tasks?

Reply:
I would think so, but is it necessary here for any of the states? I think it makes sense to get the cost data and put it into the metadata, but I'm not sure if it's necessary to add it here when we get the status. I'm going to remove it from all occurrences for now.

Reply:
Ah, I think I get it - these are terminal states, so they don't need to carry cost data with them?

Reply:
Yeah, I wouldn't think the terminal states would care about whether or not we fetched the cost data since we'll stop polling anyway.